capstone/src/main/java/com/google/sps/servlets/ChallengeServlet.java

```diff
@@ -18,47 +15,46 @@
 import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.appengine.repackaged.com.google.common.collect.Iterables;
 import com.google.sps.Objects.Challenge;
+import com.google.sps.Objects.Time;
+import com.google.sps.Objects.response.ChallengeResponse;
 
 @WebServlet("challenge")
-public class ChallengeServlet extends HttpServlet {
+public class ChallengeServlet extends AuthenticatedServlet {
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doGet(String userId, HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     Query query = new Query("Challenge").addSort("timestamp", SortDirection.DESCENDING);
     PreparedQuery results = datastore.prepare(query);
     Challenge challenge = null;
+    ChallengeResponse challengeResponse = null;
     // Check if there are challenges in database
     if (Iterables.size(results.asIterable()) > 0) {
       // Get most recent challenge in database
       Entity entity = results.asIterable().iterator().next();
       challenge = Challenge.fromEntity(entity);
+      // Gets whether current user has completed challenge
+      boolean hasUserCompleted = challenge.getHasUserCompleted(userId);
+      challengeResponse = new ChallengeResponse(challenge, hasUserCompleted);
     }
-    ServletHelper.write(response, challenge, "application/json");
+    ServletHelper.write(response, challengeResponse, "application/json");
   }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doPost(String userId, HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
     String challengeName = request.getParameter("name");
-    LocalDateTime dueDate = this.getDueDate(LocalDateTime.now());
-    long dueDateMillis = Timestamp.valueOf(dueDate).getTime();
+    Time time = new Time();
+    long dueDate = Challenge.getDueDate(time);
```

Comment:
nit: inline the Time instantiation instead of assigning it to a variable that you only use here